cmd/dockerd/daemon_windows.go

```diff
@@ -93,7 +94,30 @@ func newCgroupParent(config *config.Config) string {
 	return ""
 }
 
-func (cli *DaemonCli) initContainerD(_ context.Context) (func(time.Duration) error, error) {
+func (cli *DaemonCli) initContainerD(ctx context.Context) (func(time.Duration) error, error) {
+	if cli.Config.ContainerdAddr == "" && cli.Config.DefaultRuntime == config.WindowsV2RuntimeName {
+		cli.Config.ContainerdAddr = supervisor.GrpcPipeName
+		system.InitContainerdRuntime(cli.Config.ContainerdAddr)
+
+		var waitForShutdown func(time.Duration) error
+		logrus.Debug("Starting daemon managed containerd")
+		opts, err := cli.getContainerdDaemonOpts()
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to generate containerd options")
+		}
+
+		r, err := supervisor.Start(ctx, filepath.Join(cli.Config.Root, "containerd"), filepath.Join(cli.Config.ExecRoot, "containerd"), opts...)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to start containerd")
+		}
+		logrus.Debug("Started daemon managed containerd")
+		cli.Config.ContainerdAddr = r.Address()
+
+		// Try to wait for containerd to shutdown
+		waitForShutdown = r.WaitTimeout
+		return waitForShutdown, nil
```

Comment:
`waitForShutdown` is only used here, so this could be inlined (and the var further up removed);

```go
return r.WaitTimeout, nil
```

Reply:
With my suggestions applied, this function would look like;

```go
func (cli *DaemonCli) initContainerd(ctx context.Context) (func(time.Duration) error, error) {
	if cli.ContainerdAddr != "" {
		// use system containerd at the given address.
		system.InitContainerdRuntime(cli.ContainerdAddr)
		return nil, nil
	}

	if cli.DefaultRuntime != config.WindowsV2RuntimeName {
		return nil, nil
	}

	logrus.Info("containerd not running, starting managed containerd")
	opts, err := cli.getContainerdDaemonOpts()
	if err != nil {
		return nil, errors.Wrap(err, "failed to generate containerd options")
	}

	r, err := supervisor.Start(ctx, filepath.Join(cli.Root, "containerd"), filepath.Join(cli.ExecRoot, "containerd"), opts...)
	if err != nil {
		return nil, errors.Wrap(err, "failed to start containerd")
	}
	cli.ContainerdAddr = r.Address()
	system.InitContainerdRuntime(r.Address())

	// Try to wait for containerd to shutdown
	return r.WaitTimeout, nil
}
```

And the diff between Windows and Linux getting really small (maybe even up to a level that we could abstract some of the platform specifics);

<img width="1828" alt="Screenshot 2022-08-11 at 16 48 16" src="https://user-images.githubusercontent.com/1804568/184162495-8ef0d041-3d77-4915-acd5-edf172d1ddd6.png">